### ISSUE TITLE
no control of DDL sequences for indexes, FK constraints that are mutually dependent

### ISSUE THREAD

**Issue by sqlalchemy-bot** [sqlalchemy/sqlalchemy#3442]
**Migrated issue, originally created by Michael Bayer ([@zzzeek](https://github.com/zzzeek))**

the whole system described at http://docs.sqlalchemy.org/en/rel_1_0/core/ddl.html#controlling-ddl-sequences only takes place for constraints that are rendered inline within the table definition via the _create_rule() callable checked in compiler.py.  This callable is not consulted anywhere in ddl.py when it goes to create indexes, foreign keys, or for that matter sequences.  

e.g. note below they are all execute_if(postgresql) but the engine is MySQL, and they all get built

```
from sqlalchemy import *
from sqlalchemy.orm import *
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()


class A(Base):
    __tablename__ = 'a'
    id = Column(Integer, primary_key=True)
    x = Column(Integer)
    favorite_b = Column(Integer)


class B(Base):
    __tablename__ = 'b'
    id = Column(Integer, primary_key=True)
    a_id = Column(Integer)

from sqlalchemy.schema import CreateIndex, AddConstraint

idx = Index('aidx', A.x)

fk1 = ForeignKeyConstraint([B.a_id], [A.id], name='afk')
fk2 = ForeignKeyConstraint([A.favorite_b], [B.id], name='bfk')

from sqlalchemy import event

event.listen(
    Base.metadata,
    'after_create',
    CreateIndex(idx).execute_if(dialect='postgresql')
)

event.listen(
    Base.metadata,
    'after_create',
    AddConstraint(fk1).execute_if(dialect='postgresql')
)
event.listen(
    Base.metadata,
    'after_create',
    AddConstraint(fk2).execute_if(dialect='postgresql')
)

e = create_engine("mysql://scott:tiger@localhost/test", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)


```

----------------------------------------
Attachments: [3442.patch](../wiki/imported_issue_attachments/3442/3442.patch)

**Comment by zzzeek** [comment]
this is not being asked for and while it's a pretty bad inconsistency, it is at least documented for now

**Comment by zzzeek** [comment]
dupe of #7631 , fixed in 57877461c1bd3b43a9d833fbca873d59d new docs at https://docs.sqlalchemy.org/en/20/core/ddl.html#schema-ddl-ddl-if  

example of the above script:

```py
from sqlalchemy import *
from sqlalchemy.orm import *
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()


class A(Base):
    __tablename__ = 'a'
    id = Column(Integer, primary_key=True)
    x = Column(Integer)
    favorite_b = Column(Integer)


class B(Base):
    __tablename__ = 'b'
    id = Column(Integer, primary_key=True)
    a_id = Column(Integer)

from sqlalchemy.schema import CreateIndex, AddConstraint

idx = Index('aidx', A.x).ddl_if(dialect='postgresql')

fk1 = ForeignKeyConstraint([B.a_id], [A.id], name='afk').ddl_if(dialect='postgresql')
fk2 = ForeignKeyConstraint([A.favorite_b], [B.id], name='bfk').ddl_if(dialect='postgresql')


e = create_engine("mysql://scott:tiger@localhost/test", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)


```
